js/spreadsheet.js

```diff
@@ -67,248 +67,131 @@ var spreadsheet = function(options) {
-        if (allData[i][firstCol] === null) {
-          firstRow = i;
-          break;
-        }
-      }
-
-      firstRow = firstRow || 0;
-
-      // Looking for the last row in the array of allData
-      // When we got a null value in the cell it means that we reached the range borders
-      for (var i = firstRow; i < allData.length; i++) {
-        if (allData[i][firstCol] === null) {
-          lastRow = i;
-          break;
-        }
-      }
-
-      lastRow = _.max([lastRow - 1, 0]);
-    } else if (moveTo.right) {
-      // When data located on the right of the selected cell
-      firstCol = startAt[1];
-
-      for (var i = firstCol + 1; i < allData.length; i++) {
-        if (allData[startAt[0]][i] === null) {
-          lastCol = i - 1;
-          break;
-        }
-      }
+  function getDataCoordinates(tableData, cellPosition, processedCells) {
+    if (!processedCells) {
+      processedCells = [];
+      processedCells[cellPosition.x] = [];
+      processedCells[cellPosition.x][cellPosition.y] = true;
+    }
 
-      for (var i = startAt[0]; i > 0; i--) {
-        if (allData[i][lastCol] === null) {
-          firstRow = i ? i - 1 : i;
-        }
-      }
+    var cellsGroup = getNearestCells(cellPosition);
 
-      firstRow = firstRow || 0;
+    cellsGroup['currentCell'] = cellPosition;
 
-      for (var i = firstRow; i < allData.length; i++) {
-        if (allData[i][lastCol] === null) {
-          lastRow = i - 1;
-          break;
-        }
-      }
-    } else if (moveTo.top) {
-      // When data located on the top of the selected cell
-      lastRow = startAt[0];
+    for (var cell in cellsGroup) {
+      var hasData = tableData[cellsGroup[cell].x][cellsGroup[cell].y] !== null;
+      var processedRow = processedCells[cellsGroup[cell].x];
 
-      for (var i = lastRow - 1; i > 0; i--) {
-        if (allData[i][startAt[1]] === null) {
-          firstRow = i;
-          break;
-        }
+      if (!processedRow) {
+        processedCells[cellsGroup[cell].x] = [];
       }
 
-      firstRow = firstRow || 0;
+      var processedCell = processedCells[cellsGroup[cell].x][cellsGroup[cell].y];
 
-      for (var i = startAt[1]; i > 0; i--) {
-        if (allData[firstRow][i] === null) {
-          firstCol = i ? i + 1 : i;
-          break;
-        }
+      if (!processedCell) {
+        processedCells[cellsGroup[cell].x][cellsGroup[cell].y] = false;
       }
 
-      firstCol = firstCol || 0;
-
-      for (var i = firstCol; i < allData.length; i++) {
-        if (allData[firstRow][i] === null) {
-          lastCol = i - 1;
-          break;
-        }
-      }
-    } else if (moveTo.bottom) {
-      // When data located on the bottom of the selected cell
-      firstRow = startAt[0];
+      if (hasData && (!processedRow || !processedCell)) {
+        processedCells[cellsGroup[cell].x][cellsGroup[cell].y] = true;
+        getDataCoordinates(tableData, cellsGroup[cell], processedCells);
```

Comment:
Am I right in assuming that the process of selecting all cells is a recursive process, as it's indicated in this line?

If so, how efficient is it if we want to select all 100K rows and 100 columns?

Reply:
@tonytlwu You were right about large amount of data. This couldn't handle it. I've created a new [PR](https://github.com/Fliplet/fliplet-widget-data-source/pull/78) with a new solution and tested it on a large data source 60k rows.

And I'll close this PR because it is not needed anymore.